planner/core/logical_plan_builder.go

```diff
@@ -3575,7 +3641,53 @@ func getStatsTable(ctx sessionctx.Context, tblInfo *model.TableInfo, pid int64)
 func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, asName *model.CIStr) (LogicalPlan, error) {
 	dbName := tn.Schema
 	sessionVars := b.ctx.GetSessionVars()
+
 	if dbName.L == "" {
+		// Try CTE
+		for i := len(b.outerCTEs) - 1; i >= 0; i-- {
+			cte := b.outerCTEs[i]
+			if cte.def.Name.L == tn.Name.L {
+				if cte.isBuilding {
+					if cte.nonRecursive {
+						// Can't see this CTE, try outer definition.
+						continue
+					}
+
+					// Building the recursive part.
+					cte.useRecursive = true
+					if cte.seedLP == nil {
+						return nil, ErrCTERecursiveRequiresNonRecursiveFirst.FastGenByArgs(tn.Name.String())
+					}
+
+					if cte.enterSubquery || cte.recursiveRef {
+						return nil, ErrInvalidRequiresSingleReference.FastGenByArgs(tn.Name.String())
+					}
+
+					cte.recursiveRef = true
+					p := LogicalCTETable{name: cte.def.Name.String(), idForStorage: cte.storageID}.Init(b.ctx, b.getSelectOffset())
+					p.SetSchema(getResultCTESchema(cte.seedLP.Schema(), b.ctx.GetSessionVars()))
+					p.SetOutputNames(cte.seedLP.OutputNames())
+					return p, nil
+				}
+
+				b.handleHelper.pushMap(nil)
+				var p LogicalPlan
+				lp := LogicalCTE{cte: &CTEClass{IsDistinct: cte.isDistinct, seedPartLogicalPlan: cte.seedLP, recursivePartLogicalPlan: cte.recurLP, IDForStorage: cte.storageID, optFlag: cte.optFlag}}.Init(b.ctx, b.getSelectOffset())
+				lp.SetSchema(getResultCTESchema(cte.seedLP.Schema(), b.ctx.GetSessionVars()))
+				p = lp
+				p.SetOutputNames(cte.seedLP.OutputNames())
+				if len(asName.String()) > 0 {
+					var on types.NameSlice
+					for i, name := range p.OutputNames() {
+						cpOn := *name
+						on = append(on, &cpOn)
+						on[i].TblName = *asName
+					}
```

Comment:
```suggestion
					for _, name := range p.OutputNames() {
						cpOn := *name
						cpOn.TblName = *asName
						on = append(on, &cpOn)
					}
```
I prefer this, it is easier to understand.